### PR TITLE
lnrpc+lncli: adds required wallet reserve rpc & cli

### LINKED ISSUE #5540: Feature: RPC method to compute required wallet reserve
### Background

I'd like to report wallet reserve in the UI of my app so that the user is notified if he attempts to deposit less than the required amount. I can of course compute it myself but then I have to worry about the constants changing and updating the app with a new LND release. Having the algorithm exposed in form of RPC method would improve reliability.

### Proposed RPC method

Method name: `GetRequiredWalletReserve`
Input argument: `additionalPublicChannels: u32` - how many *additional* channels the caller would like to open
Output fields:
* `requiredReserve: BitcoinAmount` - the result, currently `min((currentPublicChannels + additionalPublicChannels) * 10000, 100000)`
* `walletBalance: BitcoinAmount` - a small optimization so that the caller doesn't have to make two RPC calls

### Alternatives

Stabilize the current algorithm (promise not to change it).

### PR DESCRIPTION
**Pull request by priyanshiiit**
## Change Description
Fixes #5540 

Adds rpc and cli command for required wallet reserve.

Screenshot from 2022-05-29 21-24-37

## Steps to Test

```
1. make rpc
2. make install tags="walletrpc"
3. lncli wallet requiredreserve
```

### Testing
- [ ] Your PR passes all CI checks.
- [ ] Tests covering the positive and negative (error paths) are included.
- [ ] Bug fixes contain tests triggering the bug to prevent regressions.

### Code Style and Documentation
- [x] The change obeys the [Code Documentation and Commenting](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#CodeDocumentation) guidelines, and lines wrap at 80.
- [x] Commits follow the [Ideal Git Commit Structure](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md#IdealGitCommitStructure).
- [ ] Any new logging statements use an appropriate subsystem and logging level.
- [ ]  [There is a change description in the release notes](https://github.com/lightningnetwork/lnd/tree/master/docs/release-notes), or `[skip ci]` in the commit message for small changes.

📝 Please see our  [Contribution Guidelines](https://github.com/lightningnetwork/lnd/blob/master/docs/code_contribution_guidelines.md) for further guidance.

### CODE REVIEW

File: lnrpc/walletrpc/config_active.go [hunk 1]
```diff
@@ -44,7 +44,7 @@ type Config struct {
 
 	// Wallet is the primary wallet that the WalletKit will use to proxy
 	// any relevant requests to.
-	Wallet lnwallet.WalletController
+	Wallet *lnwallet.LightningWallet
```
Comment (yyforyongyu) [inline lnrpc/walletrpc/config_active.go:47]: Why this change? I think the interface needs to be left here so it's easier to write unit tests. Can we add the method `CurrentNumAnchorChans` to that interface and use it here?

File: lnrpc/walletrpc/walletkit_server.go [hunk 2]
```diff
@@ -40,6 +40,12 @@ import (
 	"gopkg.in/macaroon-bakery.v2/bakery"
 )
 
+const (
+	// anchorChanReservedValue is the amount we'll keep around in the
+	// wallet in case we have to fee bump anchor channels on force close.
+	anchorChanReservedValue = btcutil.Amount(10_000)
```
Comment (yyforyongyu) [inline lnrpc/walletrpc/walletkit_server.go:46]: We can also export this value from `lnwallet` to avoid dup definitions.

File: lnrpc/walletrpc/walletkit_server.go [hunk 3]
```diff
@@ -1448,6 +1458,39 @@ func (w *WalletKit) ListAccounts(ctx context.Context,
 	return &ListAccountsResponse{Accounts: rpcAccounts}, nil
 }
 
+// RequiredReserve returns the required reserve for a wallet along
+// with its balance.
+func (w *WalletKit) RequiredReserve(ctx context.Context,
```
Comment (yyforyongyu) [inline lnrpc/walletrpc/walletkit_server.go:1463]: Missing test.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Thanks for the PR! My major comment is whether we should replace the interface, and the new rpc method is missing tests.

**Comment by priyanshiiit** [comment]
Hi @yyforyongyu! Thanks so much for the review!

I have updated the PR by moving `CurrentNumAnchorChans` to `WalletController` interface. Please let me know if I'm implementing it in the right direction.

PS: I'll update the commit message and add tests as well. Just wanted to check the approach for now.

File: lnrpc/walletrpc/config_active.go (hunk 1 above)
Comment (priyanshiiit) [reply lnrpc/walletrpc/config_active.go]: I think the `WalletController` interface is for non Lightning Network specifications, not sure if it be fine to add  `CurrentNumAnchorChans` there.

**Review by carlaKC — COMMENTED** [review]
Drive by comment, but wouldn't this be a good candidate for adding a new field to [WalletBalanceResponse](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L2539) rather than a new api?

File: lnwallet/btcwallet/btcwallet.go [hunk 4]
```diff
@@ -710,6 +712,60 @@ func (b *BtcWallet) ListAccounts(name string,
 	return res, nil
 }
 
+// CurrentNumAnchorChans returns the current number of non-private anchor
+// channels the wallet should be ready to fee bump if needed.
+func (b *BtcWallet) CurrentNumAnchorChans() (int, error) {
```
Comment (guggero) [inline lnwallet/btcwallet/btcwallet.go:717]: Hmm, I don't really like this code move since it requires the btc wallet to know about the channel DB, which is a clear violation of the responsibility boundaries...
I think the better alternative is probably to add the `CurrentNumAnchorChans()` as a function argument to the `lnrpc/walletrpc/config_active.go` file. See the example in `lnrpc/peersrpc/config_active.go`. That probably results in the smallest diff.

File: lnwallet/wallet.go [hunk 5]
```diff
@@ -35,18 +35,18 @@ const (
 	// outside word.
 	msgBufferSize = 100
 
-	// anchorChanReservedValue is the amount we'll keep around in the
+	// AnchorChanReservedValue is the amount we'll keep around in the
 	// wallet in case we have to fee bump anchor channels on force close.
 	// TODO(halseth): update constant to target a specific commit size at
 	// set fee rate.
-	anchorChanReservedValue = btcutil.Amount(10_000)
+	AnchorChanReservedValue = btcutil.Amount(10_000)
```
Comment (guggero) [inline lnwallet/wallet.go:42]: :+1:

File: cmd/lncli/walletrpc_active.go [hunk 6]
```diff
@@ -1083,6 +1084,50 @@ func listAccounts(ctx *cli.Context) error {
...
+	Usage: "Returns the wallet reserve ",
+	Description: `
+	Returns the required reserve for a wallet along with its balance.
+
+	Use the flag --additionalChannels to get the reserve value based
+	on the additional channels you would like to open.
+	`,
+	Flags: []cli.Flag{
+		cli.Uint64Flag{
+			Name: "additionalChannels",
```
Comment (guggero) [inline cmd/lncli/walletrpc_active.go:1098]: nit: we normally use `snake_case` for command line arguments, not camelCase.

File: lnrpc/walletrpc/walletkit.proto [hunk 7]
```diff
@@ -400,6 +407,19 @@ message ListAccountsResponse {
...
+    // The number of additional channels the user would like to open.
+    uint32 additional_public_channels = 1;
+}
+
+message RequiredReserveResponse {
+    // The amount of reserve required.
+    int64 required_reserve = 1;
+
+    // The balance of the wallet.
+    int64 wallet_balance = 2;
```
Comment (guggero) [inline lnrpc/walletrpc/walletkit.proto:420]: I honestly also don't see the point of duplicating the wallet balance again here. I think we should remove it to have a concise API.

**Review by guggero — COMMENTED** [review]
Thanks for taking a stab at this!
Looking back at [my comment](https://github.com/lightningnetwork/lnd/issues/5540#issuecomment-1074999652), it was probably bad advice, seeing @yyforyongyu's arguments for testability. See my inline comment about the third option, which is probably the easiest to go with.

**Comment by guggero** [comment]
> Drive by comment, but wouldn't this be a good candidate for adding a new field to [WalletBalanceResponse](https://github.com/lightningnetwork/lnd/blob/master/lnrpc/lightning.proto#L2539) rather than a new api?

While I think that makes sense, it would not allow us to add the `additional_channels` parameter (as that definitely would look out of place in the wallet balance request). But maybe we can also add the current required reserve to the wallet balance and keep the extra RPC for the `additional_channels` flag?

File: lnrpc/walletrpc/walletkit_server.go [hunk 8]
```diff
@@ -1448,6 +1452,26 @@ func (w *WalletKit) ListAccounts(ctx context.Context,
...
+	req *RequiredReserveRequest) (*RequiredReserveResponse, error) {
+
+	numAnchorChans, err := w.cfg.CurrentNumAnchorChans()
+	anchorChanReservedValue := lnwallet.AnchorChanReservedValue
```
Comment (guggero) [inline lnrpc/walletrpc/walletkit_server.go:1461]: nit: move this variable declaration (`anchorChanReservedValue`) further down to not break the visual pattern of function/method call followed by the error check.

File: rpcserver.go [hunk 9]
```diff
@@ -750,11 +750,11 @@ func (r *rpcServer) addDeps(s *server, macService *macaroons.Service,
 	// TODO(roasbeef): extend sub-sever config to have both (local vs remote) DB
 	err = subServerCgs.PopulateDependencies(
 		r.cfg, s.cc, r.cfg.networkDir, macService, atpl, invoiceRegistry,
-		s.htlcSwitch, r.cfg.ActiveNetParams.Params, s.chanRouter,
-		routerBackend, s.nodeSigner, s.graphDB, s.chanStateDB,
-		s.sweeper, tower, s.towerClient, s.anchorTowerClient,
-		r.cfg.net.ResolveTCPAddr, genInvoiceFeatures,
-		genAmpInvoiceFeatures, getNodeAnnouncement,
+		s.htlcSwitch, r.cfg.ActiveNetParams.Params,
+		s.cc.Wallet.CurrentNumAnchorChans, s.chanRouter, routerBackend,
```
Comment (guggero) [inline rpcserver.go:754]: We don't need to pass this function in here specifically since we already pass in `s.cc`, we can directly use `cc.Wallet.CurrentNumAnchorChans` in `subserver_config.go`.

**Comment by guggero** [comment]
Ah, I forgot. Didn't we also discuss adding the current reserve to the output of `WalletBalance`?

**Comment by priyanshiiit** [comment]
> Ah, I forgot. Didn't we also discuss adding the current reserve to the output of `WalletBalance`?

Ahh, yes I also forgot. Added now :blush:

File: cmd/lncli/walletrpc_active.go [hunk 10]
```diff
@@ -1083,6 +1084,50 @@ func listAccounts(ctx *cli.Context) error {
 	return nil
 }
 
+var requiredReserveCommand = cli.Command{
+	Name:  "requiredreserve",
+	Usage: "Returns the wallet reserve ",
```
Comment (guggero) [inline cmd/lncli/walletrpc_active.go:1089]: nit: extra space at the end, should be a full stop instead.

File: cmd/lncli/walletrpc_active.go [hunk 11]
```diff
@@ -1083,6 +1084,50 @@ func listAccounts(ctx *cli.Context) error {
...
+	Description: `
+	Returns the required reserve for a wallet along with its balance.
+
+	Use the flag --additional_channels to get the reserve value based
+	on the additional channels you would like to open.
+	`,
+	Flags: []cli.Flag{
+		cli.Uint64Flag{
+			Name: "additional_channels",
+			Usage: "(optional) specify the additional public channels" +
```
Comment (guggero) [inline cmd/lncli/walletrpc_active.go:1099]: Space missing in string before line break.

File: cmd/lncli/walletrpc_active.go [hunk 12]
```diff
@@ -1083,6 +1084,50 @@ func listAccounts(ctx *cli.Context) error {
 	return nil
 }
 
+var requiredReserveCommand = cli.Command{
+	Name:  "requiredreserve",
+	Usage: "Returns the wallet reserve ",
+	Description: `
+	Returns the required reserve for a wallet along with its balance.
```
Comment (guggero) [inline cmd/lncli/walletrpc_active.go:1091]: Should we give some more context here what the reserve is for?
For example:

```
The required reserve is the minimum amount of satoshis that should be kept in
the wallet in order to fee bump anchor channels if necessary. The value scales
with the number of public anchor channels but is capped at a maximum.
```

File: lnrpc/walletrpc/walletkit.proto [hunk 13]
```diff
@@ -64,6 +64,13 @@ service WalletKit {
     */
     rpc ListAccounts (ListAccountsRequest) returns (ListAccountsResponse);
 
+    /*
+    RequiredReserve returns the required reserve for a wallet along with its
+    balance.
```
Comment (guggero) [inline lnrpc/walletrpc/walletkit.proto:69]: This could also use some more information/context. Feel free to use the same text I commented above.

File: lnrpc/lightning.proto [hunk 14]
```diff
@@ -2550,6 +2550,9 @@ message WalletBalanceResponse {
     // other usage.
     int64 locked_balance = 5;
 
+    // The amount of reserve required.
+    int64 required_reserve = 6;
```
Comment (guggero) [inline lnrpc/lightning.proto]: This change shouldn't be in this commit yet since it's changed again later.

File: lnrpc/walletrpc/walletkit_server.go [hunk 15]
```diff
@@ -1448,6 +1452,25 @@ func (w *WalletKit) ListAccounts(ctx context.Context,
...
+// RequiredReserve returns the required reserve for a wallet.
+func (w *WalletKit) RequiredReserve(ctx context.Context,
+	req *RequiredReserveRequest) (*RequiredReserveResponse, error) {
+
+	numAnchorChans, err := w.cfg.CurrentNumAnchorChans()
+	if err != nil {
+		return nil, err
+	}
+
+	anchorChanReservedValue := lnwallet.AnchorChanReservedValue
```
Comment (guggero) [inline lnrpc/walletrpc/walletkit_server.go:1464]: If you re-order the commits and add the wallet balance changes first, you could simply call `w.cfg.Wallet.RequiredReserve(totalChans)` here.

File: lnwallet/btcwallet/btcwallet.go [hunk 16]
```diff
@@ -710,6 +710,16 @@ func (b *BtcWallet) ListAccounts(name string,
 	return res, nil
 }
 
+// RequiredReserve returns the required reserve for a wallet.
+func (b *BtcWallet) RequiredReserve(
+	numAnchorChans int) btcutil.Amount {
+
+	anchorChanReservedValue := lnwallet.AnchorChanReservedValue
+	reserved := btcutil.Amount(numAnchorChans) * anchorChanReservedValue
```
Comment (guggero) [inline lnwallet/btcwallet/btcwallet.go:718]: This value should be capped at `lnwallet.maxAnchorChanReservedValue` (needs to be exported as well).

File: lntest/mock/walletcontroller.go [hunk 17]
```diff
@@ -105,6 +105,11 @@ func (w *WalletController) ListAccounts(string,
 	return nil, nil
 }
 
+// RequiredReserve currently returns a dummy value.
+func (w *WalletController) RequiredReserve(int) btcutil.Amount {
```
Comment (guggero) [inline lntest/mock/walletcontroller.go:109]: This change needs to be in the same commit where you add the method to the interface, otherwise that commit won't compile on its own.

File: lnwallet/interface.go [hunk 18]
```diff
@@ -230,6 +230,9 @@ type WalletController interface {
 	ListAccounts(string, *waddrmgr.KeyScope) ([]*waddrmgr.AccountProperties,
 		error)
 
+	// RequiredReserve returns the required reserve for a wallet.
```
[the remark this answers is not in the document]
Comment (priyanshiiit) [reply lnwallet/interface.go:233]: I updated all such occurrences with:

> RequiredReserve returns the minimum amount of satoshis that should be kept in
> the wallet in order to fee bump anchor channels if necessary. The value scales
> with the number of public anchor channels but is capped at a maximum.

Should I switch to this one?

File: rpcserver.go [hunk 19]
```diff
@@ -3136,16 +3136,26 @@ func (r *rpcServer) WalletBalance(ctx context.Context,
...
-		TotalBalance:       int64(totalBalance),
-		ConfirmedBalance:   int64(confirmedBalance),
-		UnconfirmedBalance: int64(unconfirmedBalance),
-		LockedBalance:      int64(lockedBalance),
-		AccountBalance:     rpcAccountBalances,
+		TotalBalance:              int64(totalBalance),
+		ConfirmedBalance:          int64(confirmedBalance),
+		UnconfirmedBalance:        int64(unconfirmedBalance),
+		LockedBalance:             int64(lockedBalance),
+		ReservedBalanceAnchorChan: int64(requiredReserve),
```
Comment (yyforyongyu) [inline rpcserver.go:3157]: Ideally whenever a new field is added we'd need to test it in our itest. For now I think we could check it in `testAnchorReservedValue`.

**Review by yyforyongyu — CHANGES_REQUESTED** [review]
Cool I think this is close! Commits are easy to follow and also a nice workaround re `CurrentNumAnchorChans`. No major comments, just mo tests on the newly added rpc field and method.

File: lntest/itest/lnd_onchain_test.go [hunk 20]
```diff
@@ -341,6 +341,40 @@ func testAnchorReservedValue(net *lntest.NetworkHarness, t *harnessTest) {
...
 	aliceBalance := waitForConfirmedBalance()
 
+	// The reserved value should be equal to the required reserve for anchor
+	// channels.
+	walletBalanceResp, err := alice.WalletBalance(
+		ctxb, &lnrpc.WalletBalanceRequest{},
+	)
+	require.NoError(t.t, err)
+	require.Equal(
+		t.t, walletBalanceResp.ReservedBalanceAnchorChan, aliceBalance,
```
Comment (guggero) [inline lntest/itest/lnd_onchain_test.go:351]: nit: I think in `Equal` the first parameter (after `t`) is always the expected value and the second parameter is the value to test. So I think this should be swapped for an eventual error message to be correct?
Same with the `Equal()` below.

**Review by guggero — APPROVED** [review]
Great PR, almost spotless. Amazing work, LGTM :tada:

File: docs/release-notes/release-notes-0.15.1.md [hunk 21]
```diff
@@ -15,8 +15,14 @@
...
+## RPC Server
+
+* [Add wallet reserve rpc & field in wallet balance](https://github.com/lightningnetwork/lnd/pull/6592)
+
 # Contributors (Alphabetical Order)
 
 * Elle Mouton
 * ErikEk
 * Priyansh Rastogi
+
```
Comment (yyforyongyu) [inline docs/release-notes/release-notes-0.15.1.md:27]: nit: extra newlines

**Comment by guggero** [comment]
Needs another rebase.

**Comment by priyanshiiit** [comment]
Thanks for the final review⚡Updated!
